### PR TITLE
Minimum working for dev server

### PR DESCRIPTION
**Pull request by ucamrc**
(For the record this branch has been tested on EC2 (m3.medium / HVM / SSD) and the core provisioning time is less than 10 minutes! However, the remaining "non-core" time is spent importing timetable data, you need to add that as extra time to get the total time and I have been testing with a reduced dataset; tomorrow I will benchmark how long it takes to import the full dataset.)

### CODE REVIEW

File: modules/ghservice/templates/apache/app_admin.conf.erb [hunk 1]
```diff
@@ -0,0 +1,95 @@
+    ##
```
Comment (bertpareyn) [inline modules/ghservice/templates/apache/app_admin.conf.erb:1]: All extras or changes in these configuration files should be put in a PR to grasshopper-ui as well so we don't lose track of them.

File: modules/ghservice/manifests/apache.pp [hunk 2]
```diff
@@ -1,6 +1,87 @@
...
+
+    $apache_dir_require = $enable_basic_auth ? {
+        'true'  => 'valid-user',
+        default => 'all granted',
+    }
+
+    $path_robotstxt         = '/var/www/robots.txt'
+
+    file { $path_robotstxt:
+       source => 'puppet:///modules/ghservice/robots.txt.disallowall'
```
Comment (Coenego) [inline modules/ghservice/manifests/apache.pp:38]: One extra space needed for indention

File: README.md [hunk 3]
```diff
@@ -8,6 +8,61 @@ Puppet configuration and environment management for the Grasshopper event engine
...
+local$ scp timetabledata.json devserver.ontheinternet:/tmp/timetabledata.json
+
+# Back to the server to run puppet
+# This next step could take up to around 60mins if you have a slow server and lots of data to import!
+# Take note of the puppet command line displayed at the end, you may find it useful later
+devserver$ sudo ./provisioning/grasshopper/init.sh
+
+# Set username and password to protect externally visible server with play data in
+# (This only applies in environments where ghservice::apache::enable_basic_auth is true)
+devserver$ sudo htpasswd -c /etc/apache2/dev_auth_file #username#
```
Comment (bertpareyn) [inline README.md:38]: Can we leave this out? I don't think we should be overprotecting this.

File: README.md [hunk 4]
```diff
@@ -8,6 +8,61 @@ Puppet configuration and environment management for the Grasshopper event engine
...
+# - to change git config to match a tag if appropriate
+devserver$ sudo vim environments/dev/hiera/common.json
+
+# Copy some timetable data to import onto the server
+local$ scp timetabledata.json devserver.ontheinternet:/tmp/timetabledata.json
+
+# Back to the server to run puppet
+# This next step could take up to around 60mins if you have a slow server and lots of data to import!
+# Take note of the puppet command line displayed at the end, you may find it useful later
+devserver$ sudo ./provisioning/grasshopper/init.sh
```
Comment (bertpareyn) [inline README.md:34]: Can all steps prior to this be puppetised as well? It'd be nice to only have to run a single command.

File: README.md [hunk 5]
```diff
@@ -8,6 +8,61 @@ Puppet configuration and environment management for the Grasshopper event engine
  * If you make changes to the backend code you will need to restart the app server. This can be done by ssh'ing into the client machine by running `service grasshopper restart`.
  * Even if you'd install all the components on your host OS, you would not be able to run the server as some of the npm modules are compiled during the provisioning step.
 
+### Dev Environment (e.g. on EC2)
+
+```
```
Comment (bertpareyn) [inline README.md:13]: Minor: could you take all of this out of the comments block and only use `this style of markup` for the lines to execute?

File: README.md [hunk 6]
```diff
@@ -8,6 +8,61 @@ Puppet configuration and environment management for the Grasshopper event engine
...
+# Provision an Ubuntu Trusty server
```
Comment (bertpareyn) [inline README.md:14]: If you have a specific AMI preference, should we add it here?

File: environments/dev/hiera/common.json [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        "::ghservice::deps::common",
+        "::ghservice::apache",
+        "::ghservice::grasshopper",
+        "::ghservice::postgresql",
+        "::ghservice::ui"
+    ],
+
+    /* Domains and hostnames */
+    "admin_hostname": "admin.%{hiera('tenant_hostname')}",
+    "tenant_hostname": "ec2-nn-nn-nn-nn.eu-west-1.compute.amazonaws.com",
```
Comment (bertpareyn) [inline environments/dev/hiera/common.json:14]: Could this be pulled in by puppet? Is there a system variable we could speak to?

Comment (bertpareyn) [inline environments/dev/hiera/common.json:14]: Alternatively, we could specify it when running the init script

File: environments/dev/hiera/common.json [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+    /* PostgreSQL settings */
+    "ghservice::postgresql::version": "",
+    "ghservice::postgresql::db": "grasshopper",
+    "ghservice::postgresql::user": "grasshopper",
+    "ghservice::postgresql::pass": "grasshopper",
+
+    /* Apache directives */
+    "ghservice::apache::version": "",
+    "ghservice::apache::enable_basic_auth": "true"
```
Comment (bertpareyn) [inline environments/dev/hiera/common.json:66]: This can be defaulted to false so it doesn't require a login

File: modules/ghservice/manifests/deps/nodejs.pp [hunk 9]
```diff
@@ -1,11 +1,22 @@
...
         version => $nodejs_version,
         manage_repo => true
     }
+
+    # This next section is necessary to ensure the above class
+    # is actually contained within the surrounding class
+    # because by default puppet will let that nodejs declaration float freely in
+    # the dependency graph, resulting in commands like npm install trying
+    # to run too early.
+    # https://docs.puppetlabs.com/puppet/latest/reference/lang_containment.html#anchor-pattern-containment-for-compatibility-with-puppet--340
```
Comment (bertpareyn) [inline modules/ghservice/manifests/deps/nodejs.pp:16]: As a note in general on this PR, can you clean up this documentation a bit so that all lines have equal length? Makes it all more readable

File: modules/grasshopper/manifests/setup.pp [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+  exec { 'temporarily-stop-grasshopper-for-import':
+      # if file for import exists and server is responding, stop the server
+      onlyif  => "test -f /tmp/timetabledata.json && curl --fail ${admin_test_url}",
+      creates => "/opt/timetabledata.json.imported",
+      command => 'stop grasshopper && sleep 5',
+  } ->
+  exec { 'import-timetable-data':
+      onlyif  => "test -f /tmp/timetabledata.json",
+      creates => "/opt/timetabledata.json.imported",
```
Comment (bertpareyn) [inline modules/grasshopper/manifests/setup.pp:37]: Is there any specific reason for duplicating this file before importing it?

File: provisioning/common.sh [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+# Create the hiera config
+cp ${PUPPET_HIERA} /etc/puppet/hiera.yaml
+
+# Run puppet
+
+PUPPET_EXTRA_OPTS="--verbose"
+PUPPET_CMD="puppet apply ${PUPPET_EXTRA_OPTS} --modulepath environments/${PUPPET_ENV}/modules:modules:/etc/puppet/modules --certname dev0 --environment ${PUPPET_ENV} --hiera_config /etc/puppet/hiera.yaml site.pp"
+
+echo "Applying puppet catalog. This might take a while (~30+ mins is not unreasonable)"
```
Comment (bertpareyn) [inline provisioning/common.sh:35]: I was happy to see all of it finish under 10 minutes :)

File: README.md [hunk 12]
```diff
@@ -8,6 +8,61 @@ Puppet configuration and environment management for the Grasshopper event engine
  * If you make changes to the backend code you will need to restart the app server. This can be done by ssh'ing into the client machine by running `service grasshopper restart`.
  * Even if you'd install all the components on your host OS, you would not be able to run the server as some of the npm modules are compiled during the provisioning step.
 
+### Dev Environment (e.g. on EC2)
```
Comment (Coenego) [inline README.md:11]: Can you move the `Dev Environment` block under the `Local / Vagrant`?

I would prefer to have the following order (when possible) throughout all our files/folders/.. 
- local / vagrant
- dev
- (qa / staging)
- (performance)
- (production)

File: Vagrantfile [hunk 13]
```diff
@@ -7,7 +7,7 @@ Vagrant.configure("2") do |config|
   config.vm.box = "grasshopper"
 
   # Forward http traffic
-  config.vm.network :forwarded_port, host: 8123, guest: 80
+  config.vm.network "private_network", ip: "192.168.56.123"
```
Comment (Coenego) [inline Vagrantfile:10]: Just out of curiosity; what is the reason for this change again?

File: provisioning/common.sh [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+
+# Make sure all the submodules have been pulled down
+cd ${PUPPET_REPO_DIR}
+sh bin/pull.sh
+
+# Create the hiera config
+cp ${PUPPET_HIERA} /etc/puppet/hiera.yaml
+
+# Run puppet
+
```
Comment (Coenego) [inline provisioning/common.sh:31]: This empty line can probably go

File: provisioning/common.sh [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+STATUS_CODE=$?
+if [ $STATUS_CODE -ne 0 ] ; then
+    echo "Installing curl"
+    apt-get -y install curl
+fi
+
+# Enable multiverse repositories
+echo "Enable multiverse repositories"
+sed -i "/^# deb.*multiverse/ s/^# //" /etc/apt/sources.list
+#apt-get update
```
Comment (Coenego) [inline provisioning/common.sh:21]: If we're not executing this, it can probably go

File: modules/ghservice/manifests/postgresql.pp [hunk 16]
```diff
@@ -1 +1,18 @@
-class ghservice::postgresql {}
+class ghservice::postgresql (
+        $db = 'grasshopper',
+        $user = 'grasshopper',
+        $password = 'grasshopper',
+    ) {
+
+    class { 'postgresql::server': }
+
```
Comment (Coenego) [inline modules/ghservice/manifests/postgresql.pp:8]: This empty line can probably go

File: README.md (hunk 4 above)
Comment (bertpareyn) [inline README.md:34]: When running the command again with a full dataset it's erroring out after a while with this error:

``` puppet
Notice: /Stage[main]/Grasshopper::Setup/Exec[temporarily-stop-grasshopper-for-import]/returns: executed successfully
Error: Command exceeded timeout
Wrapped exception:
execution expired
Error: /Stage[main]/Grasshopper::Setup/Exec[import-timetable-data]/returns: change from notrun to 0 failed: Command exceeded timeout
Notice: /Stage[main]/Grasshopper/Service[grasshopper]: Dependency Exec[import-timetable-data] has failures: true
Warning: /Stage[main]/Grasshopper/Service[grasshopper]: Skipping because of failed dependencies
Notice: Finished catalog run in 336.25 seconds
```

**Comment by bertpareyn** [comment]
Great work! Back to @ucamrc for followup

**Comment by Coenego** [comment]
Great work +1. Looks really good! Left a couple of small comments.

**Comment by ucamrc** [comment]
As agreed verbally, I will do the respond to the remainder of the applicable above comments in the next pull request, let's get this one merged first.

**Comment by bertpareyn** [comment]
Merging, created an issue at https://github.com/CUL-DigitalServices/grasshopper-puppet/issues/4 to follow up on review

**Comment by bertpareyn** [comment]
Something I found out after updating a node is that the grasshopper service should be restarted when changes were pulled in. Could you add that in your followup PR as well @ucamrc ?

File: modules/ghservice/templates/apache/app_admin.conf.erb (hunk 1 above)
Comment (ucamrc) [inline modules/ghservice/templates/apache/app_admin.conf.erb:1]: The only two changes to the Apache conf that I added were for robots.txt and auth-protection - do you really want these in grasshopper-ui?
